pkg/vault/test/foundry/utils/BaseExtremeAmountsTest.sol

```diff
@@ -0,0 +1,546 @@
+    }
+
+    function _calculateExactAmountsIn(uint256[2] memory amountsIn) private view returns (uint256[] memory) {
+        uint256[] memory exactAmountsIn = new uint256[](2);
+        uint256[] memory maxAmountsIn = _calculateMaxAmountIn();
+        exactAmountsIn[0] = bound(amountsIn[0], initBalances[0], maxAmountsIn[0]);
+        exactAmountsIn[1] = bound(amountsIn[1], initBalances[1], maxAmountsIn[1]);
+
+        return exactAmountsIn;
+    }
+
+    function _calculateMaxAmountIn() private view returns (uint256[] memory) {
+        uint256[] memory maxAmountsIn = new uint256[](2);
+        for (uint256 i = 0; i < maxAmountsIn.length; i++) {
+            uint256 maxAmount = initBalances[i] * 100;
+
+            // if invariant ratio is really high, then we don't need to receive an exception
+            unchecked {
+                maxAmountsIn[i] = ((initBalances[i] * maxInvariantRatio) / 1e18);
+            }
+
+            if (maxAmountsIn[i] > maxAmount || maxAmountsIn[i] < initBalances[i]) {
+                maxAmountsIn[i] = maxAmount;
+            }
+
+            maxAmountsIn[i] = maxAmountsIn[i] - initBalances[i];
+        }
+
+        return maxAmountsIn;
+    }
+
+    function _testAddUnbalancedAndRemoveLiquidityProportional(uint256[] memory exactAmountsIn) private {
+        vault.forceUnlock();
+
+        (uint256[] memory amountsIn, uint256 bptAmountOut, ) = vault.addLiquidity(
+            AddLiquidityParams({
+                pool: pool,
+                to: address(this),
+                maxAmountsIn: exactAmountsIn,
+                minBptAmountOut: 0,
+                kind: AddLiquidityKind.UNBALANCED,
+                userData: bytes("")
+            })
+        );
+        assertEq(amountsIn[0], exactAmountsIn[0], "amountsIn[0] should be equal to exactAmountsIn[0]");
+        assertEq(amountsIn[1], exactAmountsIn[1], "amountsIn[1] should be equal to exactAmountsIn[1]");
+
+        (uint256 bptAmountIn, uint256[] memory amountsOut, ) = vault.removeLiquidity(
+            RemoveLiquidityParams({
+                pool: pool,
+                from: address(this),
+                maxBptAmountIn: bptAmountOut,
+                minAmountsOut: new uint256[](2),
+                kind: RemoveLiquidityKind.PROPORTIONAL,
+                userData: bytes("")
+            })
+        );
+        assertEq(bptAmountIn, bptAmountOut, "bptAmountIn should be equal to bptAmountOut");
+
+        // if amountsOut[0] more then amountsIn[0], then amountsOut[1] should be less than amountsIn[1] otherwise user gets free tokens
+        // if amountsOut[0] equal to amountsIn[0], then amountsOut[1] should be less or equal to amountsIn[1] otherwise user gets free tokens
+        // if amountsOut[0] less then amountsIn[0], we can't check anything because user doesn't get free tokens
+        if (amountsOut[0] > amountsIn[0]) {
+            assertLt(amountsOut[1], amountsIn[1], "amountsOut[1] should be less than amountsIn[1]");
```

Comment:
This case doesn't mean that no value was extracted right?

In other words, if the assertion fails for sure there's something wrong, but if it doesn't fail there's still some room for something to be wrong.

I think this is fine for a first iteration. Let's leave a comment; we can check the invariants afterwards.

Reply:
I don't quite understand what's wrong here

Reply:
`amountsOut[0] > amountsIn[0]` means that the user got some `token0` out.

If the assertion fails it means that it also got `token1` out, which means the user got free money for sure.

But let's suppose that `amountsOut[0] >>> amountsIn[0]` (user got a lot of token0 out), and then `amountsOut[1] == amountsIn[1]` (or maybe just a bit larger, but not much). That difference of `token0` that the user got might be more valuable than what they paid for it in `token1`. If I'm not mistaken, this assertion is not catching that case.

Reply:
> If the assertion fails it means that it also got token1 out, which means the user got free money for sure. So if the assertion fails, we know for sure that something is wrong.

It means that he got token0 less and token1 more than before operations. This case is legal.

> But let's suppose that amountsOut[0] >>> amountsIn[0] (user got a lot of token0 out), and then amountsOut[1] == amountsIn[1] (or maybe just a bit larger, but not much). That difference of token0 that the user got might be more valuable than what they paid for it in token1. If I'm not mistaken, this assertion is not catching that case.

you mean if the user gets token0 more and token1 less, then the first token could be much more valuable, right?

Reply:
yes, exactly.
I think the only way to address this is computing the invariants before and after. We can do that in these tests, but there's always the issue with invariant error. The solution to that might be computing the invariant in Python with more precision, but that's out of scope for this PR.